### PR TITLE
Create a single source of truth

### PR DESCRIPTION
**Pull request by bitdivine**
# Motivation
Build configuration is fragile aand complex at the moment, with the same build parameter typically being set, or defaulting, to various values in different places.

The goal here is to define a single, ideally small, set of build parameters that work for all parts of the code.  This can then be used as a single source of truth and values can be read directly from there.

# Changes
- Define a single source of truth for build configuration.

# Tests
The values computed in this config MUST be the same as those used in mainnet.  It SHOULD be the same for local and testnet but that is negotiable.

Configuration takes effect on mainnet in these places:
- `frontend/dart`
- `frontend/ts`
- `frontend/svelte`
- `rs`

## Mainnet
Example output:
```
{
  "GOVERNANCE_CANISTER_ID": "rrkah-fqaaa-aaaaa-aaaaq-cai",
  "LEDGER_CANISTER_ID": "ryjl3-tyaaa-aaaaa-aaaba-cai",
  "BUILD": "release",
  "OWN_CANISTER_ID": "qoctq-giaaa-aaaaa-aaaea-cai",
  "FETCH_ROOT_KEY": false,
  "REDIRECT_TO_LEGACY": "prod",
  "IDENTITY_SERVICE_URL": "https://identity.ic0.app/"
}
```

#### `frontend/dart`
Dart reads config explicitly with `--dart-define` clauses.  This makes it easy to analyse:
```
# From the dart build script:
if [[ $DEPLOY_ENV = "mainnet" ]]; then
  flutter build web --web-renderer html --release --no-sound-null-safety --pwa-strategy=none --dart-define=FLUTTER_WEB_CANVASKIT_URL=/assets/canvaskit/ --dart-define=REDIRECT_TO_LEGACY="prod"
else
  # For all networks that are not mainnet, build with the staging config
  flutter build web --web-renderer html --release --no-sound-null-safety --pwa-strategy=none --dart-define=DEPLOY_ENV=staging --dart-define=REDIRECT_TO_LEGACY="${REDIRECT_TO_LEGACY:-prod}"
fi
```
The relevant variables are:
- `REDIRECT_TO_LEGACY` - this is "prod" which matches the new config.
- `BUILD` - the branch above should not, in future, branch on mainnet but rather on `BUILD`.  The only part affected by DEPLOY_ENV or BUILD is whether the canvaskit is used.

Conclusion:  The new config matches the existing one where relevant.

#### `frontend/ts`
`ts` gets its config from a JSON file created by `update_config.sh`.  Comparing the JSON there MAY be additions but MUST NOT be canges or deletions.  In addition, `HOST` is intentionally missing from the original config and should be absent in the new config.  HOST is used in HttpAgent and if missing, the agent uses the default value.

```
max@sinkpad:~/dfn/nns-dapp/branches/one-config (9:46)$ ./compare-configs mainnet
wrote config from DEPLOY_ENV 'mainnet' :
Config has been defined.  Let it never be changed.
FIN
{													{
												   >	  "BUILD": "release"
  "FETCH_ROOT_KEY": false										  "FETCH_ROOT_KEY": false
												   >	  "GOVERNANCE_CANISTER_ID": "rrkah-fqaaa-aaaaa-aaaaq-cai"
  "IDENTITY_SERVICE_URL": "https://identity.ic0.app/"							  "IDENTITY_SERVICE_URL": "https://identity.ic0.app/"
												   >	  "LEDGER_CANISTER_ID": "ryjl3-tyaaa-aaaaa-aaaba-cai"
  "OWN_CANISTER_ID": "qoctq-giaaa-aaaaa-aaaea-cai"							  "OWN_CANISTER_ID": "qoctq-giaaa-aaaaa-aaaea-cai"
												   >	  "REDIRECT_TO_LEGACY": "prod"
}													}
max@sinkpad:~/dfn/nns-dapp/branches/one-config (9:58)$ 
```

#### `frontend/svelte`
Svelte uses values read in:
- `frontend/svelte/env.config.mjs`
-  `frontend/svelte/rollup.config.js`
-  `frontend/svelte/scripts/build.index.mjs`
TODO: It might be worth concentrating svelte env readings in just one file.

The variables:
```
git grep process.env | sed -E 's/.*process.env.([a-zA-Z0-9_]+).*/\1/g' | sort | uniq
BASE_HREF
DEPLOY_ENV
FETCH_ROOT_KEY
GOVERNANCE_CANISTER_ID
HOST
IDENTITY_SERVICE_URL
LEDGER_CANISTER_ID
OWN_CANISTER_ID
REDIRECT_TO_LEGACY
ROLLUP_WATCH
```
- `BASE_HREF` - Undefined for production builds, yields "/v2/" by default.  The proposed configuration does not set this either.
- `DEPLOY_ENV` - Unaffected by this PR but of course the eventual goal is to replace this.
- `FETCH_ROOT_KEY` Set to `prod` in this PR.
- `GOVERNANCE_CANISTER_ID` - Set in the code in several places, sometimes as text, sometimes as a type.  Hard-wired to `rrkah-fqaaa-aaaaa-aaaaq-cai`, which matches the value in this PR.
- `HOST` - set to undefined for mainnet, which matches our config.
- `IDENTITY_SERVICE_URL` - for mainnet, svelte hard-wires this to https://identity.ic0.app/, which matches this PR.
- `LEDGER_CANISTER_ID` is hard-wired to `ryjl3-tyaaa-aaaaa-aaaba-cai` which matches this PR.
- `OWN_CANISTER_ID` is hard-wired in svelte, for mainnet, to `qoctq-giaaa-aaaaa-aaaea-cai`, which matches this PR.
- `REDIRECT_TO_LEGACY` is set, for mainnet, to "prod" which matches.
- `ROLLUP_WATCH` is not set in the config.

## Testnet
```
DFX_NETWORK=testnet ./config.sh 
{
  "GOVERNANCE_CANISTER_ID": "rrkah-fqaaa-aaaaa-aaaaq-cai",
  "LEDGER_CANISTER_ID": "ryjl3-tyaaa-aaaaa-aaaba-cai",
  "BUILD": "dev",
  "FETCH_ROOT_KEY": true,
  "REDIRECT_TO_LEGACY": "svelte",
  "HOST": "https://nnsdapp.dfinity.network/",
  "IDENTITY_SERVICE_URL": "https://qjdve-lqaaa-aaaaa-aaaeq-cai.nnsdapp.dfinity.network",
  "OWN_CANISTER_ID": "xxxxxx"
}
```
Looks reasonable.  I have trouble deploying to testnet at the moment so this is a bit hard to probe.

## Local
Example:
```
{
  "GOVERNANCE_CANISTER_ID": "rrkah-fqaaa-aaaaa-aaaaq-cai",
  "LEDGER_CANISTER_ID": "ryjl3-tyaaa-aaaaa-aaaba-cai",
  "OWN_CANISTER_ID": "rdmx6-jaaaa-aaaaa-aaadq-cai",
  "IDENTITY_SERVICE_URL": "http://renrk-eyaaa-aaaaa-aaada-cai.localhost:8080",
  "BUILD": "dev",
  "FETCH_ROOT_KEY": true,
  "REDIRECT_TO_LEGACY": "svelte",
  "HOST": "http://localhost:8080"
}
```

Comparison wit the ts config:
```
max@sinkpad:~/dfn/nns-dapp/branches/one-config (9:40)$ ./compare-configs local
wrote config from DEPLOY_ENV 'local' :
Config has been defined.  Let it never be changed.
FIN
{													{
												   >	  "BUILD": "dev"
  "FETCH_ROOT_KEY": true										  "FETCH_ROOT_KEY": true
												   >	  "GOVERNANCE_CANISTER_ID": "rrkah-fqaaa-aaaaa-aaaaq-cai"
  "HOST": "http://localhost:8080"									  "HOST": "http://localhost:8080"
  "IDENTITY_SERVICE_URL": ""									   |	  "IDENTITY_SERVICE_URL": "http://renrk-eyaaa-aaaaa-aaada-cai.lrenrk-eyaaa-aaaaa-aaada-cai.orenr
												   >	  "LEDGER_CANISTER_ID": "ryjl3-tyaaa-aaaaa-aaaba-cai"
  "OWN_CANISTER_ID": "rdmx6-jaaaa-aaaaa-aaadq-cai"							  "OWN_CANISTER_ID": "rdmx6-jaaaa-aaaaa-aaadq-cai"
												   >	  "REDIRECT_TO_LEGACY": "svelte"
}													}
```

### CODE REVIEW

File: update_config.sh [hunk 1]
```diff
@@ -47,7 +47,7 @@ else
 EOF
 fi
 
-echo "wrote config from DEPLOY_ENV '$DEPLOY_ENV' :"
+echo "wrote config from DEPLOY_ENV '$DEPLOY_ENV' :" >&2
```
Comment (bitdivine) [inline update_config.sh:50]: Not to be committed.

File: compare-configs [hunk 2]
```diff
@@ -0,0 +1,2 @@
+variant="${1:-mainnet}"
+diff -y -W200 <(DEPLOY_ENV="$variant" ./update_config.sh | jq --sort-keys . | sed 's/,$//g') <(DFX_NETWORK="$variant" bash config.sh | jq --sort-keys . | sed 's/,$//g')
```
Comment (bitdivine) [inline compare-configs:2]: Not to be merged.  Used to compare configurations.

File: compare-configs [hunk 3]
```diff
@@ -0,0 +1,12 @@
+#!/usr/bin/env bash
+
+# NOT TO BE MERGED INTO MAIN
```
Comment (lmuntanerDfinity) [inline compare-configs:3]: what do you mean not to be merged into main? Are you going to delete before merging this PR?

Maybe move the branch to "Draft" as well.

File: config.sh [hunk 4]
```diff
@@ -0,0 +1,80 @@
+#!/usr/bin/env bash
+set -euo pipefail
+
+##################################################
+# SINGLE SOURCE OF TRUTH FOR BUILD CONFIGURATION #
+##################################################
+
+# Usage:
+# - In bash scripts, source this file.
+# - In javascript, load config.json
```
Comment (lmuntanerDfinity) [inline config.sh:10]: This `config.sh` creates / writes in this `config.json` file, is that correct?

**Review by lmuntanerDfinity — COMMENTED** [review]
A few comments.

Am I correct is assuming that the `config.json` file will then be used but the other projects as a single source of truth? To be done in upcoming PRs?

**Review by lmuntanerDfinity — COMMENTED** [review]
Did you add "config.json" to `.gitignore`? Why isn't it in this PR?

**Comment by bitdivine** [comment]
Because the script isn't called yet but we can add it now.... done.

**Comment by bitdivine** [comment]
We may want it in dockerignore as well.  Yes, I think so.

**Pull request merged**